add git-config module

stirby opened the pull request:
Added git-config module to populate `user.email` and `user.name` with coder provided username and email
Closes #11

stirby commented:
Solves #47.

matifali commented on git-config/main.tf:
```diff
@@ -0,0 +1,36 @@
+      CODER_EMAIL = var.user_email
+    }) 
+    display_name = "Git Config"
+    icon = "/emojis/1f58c.png" 
```
```suggestion
    icon = "/icon/git.svg" 
```

stirby replied:
Thank you, didn't realize this would work. Great call!

matifali commented on git-config/main.tf:
```diff
@@ -0,0 +1,36 @@
+  description = "The email of the Coder workspace owner."
+}
+
+resource "coder_script" "personalize" {
```
```suggestion
resource "coder_script" "got_config" {
```

stirby replied:
thank you

matifali commented on git-config/README.md:
````diff
@@ -0,0 +1,40 @@
+
+```hcl
+module "git-config" {
+  source = "git::https://github.com/coder/modules.git//git-config?ref=git-config"
````
```suggestion
  source = "https://registry.coder.com/modules/git-config"
```

matifali commented on git-config/README.md:
````diff
@@ -0,0 +1,40 @@
+
+```hcl
+module "git-config" {
+  source = "git::https://github.com/coder/modules.git//git-config?ref=git-config"
````
```suggestion
  source = "git::https://github.com/coder/modules.git//git-config?ref=git-config"
```

matifali commented on git-config/README.md:
````diff
@@ -0,0 +1,40 @@
+```hcl
+module "git-config" {
+  source = "git::https://github.com/coder/modules.git//git-config?ref=git-config"
+  agent_id = coder_agent.main.id
````
```suggestion
  agent_id = coder_agent.example.id
```

matifali commented on git-config/README.md:
````diff
@@ -0,0 +1,40 @@
+```hcl
+module "git-config" {
+  source = "git::https://github.com/coder/modules.git//git-config?ref=git-config"
+  agent_id = coder_agent.main.id
````
```suggestion
  agent_id = coder_agent.example.id
```

matifali reviewed (approved):
LGTM> Thanks. Justa. few suggestions

matifali commented on git-config/main.tf:
```diff
@@ -0,0 +1,36 @@
+    display_name = "Git Config"
+    icon = "/icon/git.svg" 
+    run_on_start = true
+}
```
Can you add two coder parameters to take user_name and email from the user? If they leave it empty, we default to coder deployment username and email.?

With the current approach, Users cannot specify a custom username and email while creating the template. Only the admin can update them.

stirby replied:
Great call! Adding now.

stirby replied:
@matifali What if the parameters' values default to the owner email and username, or whatever default is set by the admin? I think this would make the operation more transparent and ergonomic.

stirby replied:
I've set the default for coder parameters to "default". If the workspace owner wants to populate the fields with thier own credentials they simply replace this value. Let me know if you think this is suitable.

matifali replied:
I think it works.

matifali reviewed (approved):
Great.

matifali commented:
Hi @stirby I updated the logic to make it simple, and I guess you were missing the main idea to use the coder email and username with an option to overwrite if needed.
I guess most devs will need to update username to a fullname

stirby commented:
> Hi @stirby, I updated the logic to make it simple, and I guess you were missing the main idea to use the coder email and username with an option to overwrite if needed. Most devs will need to update their username to a full name and not the email.
> 
> Also, I added 2 new variables so that the template admin can control what changes are allowed.
> 
> You should update the tests as needed.

I hadn't realized we could pull data from the workspace directly in the module. This is a major improvement, thank you.

matifali commented:
Can you help with the tests?

stirby commented:
I'm updating them now.

stirby commented:
@matifali I fixed the issues we encountered before. Please review the weird logic in `coder_script`. I had to use these try blocks to avoid indexing errors when `allow_email_change` or `allow_username_change` are set to false. 

Otherwise I think the ideal implementation has been achieved: The coder parameters are shown as empty when the are set to be overridden by the workspace owner credentials.

matifali reviewed (approved):
I don't quite understand the logic. But of it works good to merge.
